### PR TITLE
[ᚬframework] feat(core/binding): add service_sdk, chain_querier, request_context

### PR DESCRIPTION
**Pull request by mkxbl**
**What type of PR is this?**
feat(core/binding)
**What this PR does / why we need it**:
service_sdk: used by developer to write service of framework
chain_querier: used by sdk to query chain data
request_context: manage cycles and events
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: core/binding/src/request_context.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            caller,
+            epoch_id,
+            service_name,
+            service_method,
+            service_playload,
+            events: Rc::new(RefCell::new(Vec::new())),
+        }
+    }
+
+    pub fn derive_context(
```
Comment (yejiayu) [inline core/binding/src/request_context.rs]: with_context

File: core/binding/Cargo.toml [hunk 2]
```diff
@@ -17,3 +17,6 @@ rocksdb = "0.12"
 lazy_static = "1.4"
 byteorder = "1.3"
 rlp = "0.4"
+futures = "0.3"
+async-trait = "0.1"
```
Comment (yejiayu) [inline core/binding/Cargo.toml]: Where is the `async-trait` used?

File: core/binding/src/sdk/mod.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            chain_querier,
+            request_context,
+        }
+    }
+}
+
+impl<S: 'static + ServiceState, C: ChainQuerier, R: RequestContext> ServiceSDK
+    for DefalutServiceSDK<S, C, R>
+{
+    type ContextItem = R;
```
Comment (yejiayu) [inline core/binding/src/sdk/mod.rs]: unnecessary

File: core/binding/src/tests/request_context.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    );
+
+    ctx.sub_cycles(8).unwrap();
+    assert_eq!(ctx.get_cycles_used(), 18);
+
+    assert_eq!(ctx.get_cycles_limit(), 100);
+    assert_eq!(ctx.get_cycles_price(), 8);
+    assert_eq!(ctx.get_caller(), mock_address());
+    assert_eq!(ctx.get_current_epoch_id(), 1);
+    assert_eq!(ctx.get_service_name(), "service_name".to_owned());
```
Comment (yejiayu) [inline core/binding/src/tests/request_context.rs]: I don't think can pass the test case. 

`ctx.get_service_name()` returns `&str`
`service_name".to_owned()` returns `String`

Did you run the `cargo test`?

File: core/binding/src/request_context.rs (hunk 1 above)
Comment (mkxbl) [reply core/binding/src/request_context.rs]: fixed

File: core/binding/Cargo.toml (hunk 2 above)
Comment (mkxbl) [reply core/binding/Cargo.toml]: used in tests/sdk.rs, to mock Storage trait

File: core/binding/src/sdk/mod.rs (hunk 3 above)
Comment (mkxbl) [reply core/binding/src/sdk/mod.rs]: I think it's necessary.
```rust
pub trait ServiceSDK {
    type ContextItem: RequestContext;

```
return type of `fn get_request_context` is a generic type implementing trait RequestContext

File: core/binding/src/tests/request_context.rs (hunk 4 above)
Comment (mkxbl) [reply core/binding/src/tests/request_context.rs]: fixed.

I did run `cargo test`, but it did not complain

File: core/binding/Cargo.toml (hunk 2 above)
Comment (yejiayu) [reply core/binding/Cargo.toml]: move [devDependencies]

File: core/binding/src/sdk/mod.rs (hunk 3 above)
Comment (yejiayu) [reply core/binding/src/sdk/mod.rs]: ok, isee

**Pull request merged**